### PR TITLE
Add `bevy_hierarchy` Crate and plugin documentation 

### PR DESCRIPTION
**Pull request by Nilirad**
This PR is part of a project aimed at improving the API documentation of `bevy_hierarchy`. Other PRs will be based on this.

This PR in particular is also an experiment in providing a high level overview of the tools provided by a Bevy plugin/crate. It also provides general information about universal invariants, so statement repetition in crate items can be dramatically reduced.

## Other changes

The other PRs of this project that expand on this one:

- #10952
- #10953
- #10954
- #10955
- #10956
- #10957

### CODE REVIEW

File: crates/bevy_hierarchy/src/lib.rs [hunk 1]
```diff
@@ -1,8 +1,49 @@
...
+//! This crate introduces various tools, including a [plugin]
+//! for managing parent-child relationships between entities.
+//! It provides two components, [`Parent`] and [`Children`],
+//! to store references to related entities.
+//! It also provides [command] and [world] API extensions
+//! to set and clear those relationships.
+//!
+//! More advanced users may also appreciate
+//! [query extension methods] to traverse hierarchies,
+//! and [events] to notify hiearchical changes.
```
Comment (GitGhillie) [inline crates/bevy_hierarchy/src/lib.rs:18]:
```suggestion
//! and [events] to notify hierarchical changes.
```

File: crates/bevy_hierarchy/src/lib.rs [hunk 2]
```diff
@@ -1,8 +1,49 @@
...
+//! For example, assigning a child to an entity
+//! will always set the parent in the other,
+//! and vice versa.
+//! Similarly, unassigning a child in the parent
+//! will always unassign the parent in the child.
+//!
+//! ## Despawning entities
+//!
+//! The commands and methods provided by `bevy_ecs` to despawn entities
+//! are not capable of despawning hierarchies of entities, invalidating them.
```
Comment (GitGhillie) [inline crates/bevy_hierarchy/src/lib.rs:37]: I think you meant to say here that using the commands and methods provided by `bevy_ecs` to despawn entities can invalidate the hierarchy. Because if you despawn a hierarchy of entities there is nothing to invalidate

**Review by GitGhillie — COMMENTED** [review]
Good step forward IMO. Might be nice to link to the hierarchy example(s)

File: crates/bevy_hierarchy/src/lib.rs (hunk 2 above)
Comment (Nilirad) [reply crates/bevy_hierarchy/src/lib.rs:37]: I was saying that `bevy_ecs` items are just capable of despawning single entities. If you do that to a single entity that is part of a hierarchy, it surely gets invalidated. You'd have to manually despawn the entire hierarchy, or an entire subtree and then update the hierarchy components on the parent of the root.

I can slightly reword it to make it more clear.

**Review by GitGhillie — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Merging now, ping me on Github or on Discord for more reviews on the follow-ups as they're ready.
